chore(platform): approval flow refactoring

platon-rov opened the pull request:
#### Please provide a link to the associated issue.

Closes #5963.
Closes #5965.
Closes #4884.

#### Please provide a brief summary of this pull request.

Approval flow refactoring & bugfixes.

#### Please check whether the PR fulfills the following requirements

- [X] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/main/CONTRIBUTING.md
- [X] tests for the changes that have been done
- [X] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist
- [X] Run npm run build-pack-library and test in external application

Documentation checklist:
- [n/a] update `README.md`
- [n/a] [Breaking Changes Wiki](https://github.com/SAP/fundamental-ngx/wiki/Breaking-Changes)
- [n/a] Documentation Examples
- [X] Stackblitz works for all examples

platon-rov commented on libs/core/src/lib/grid-list/components/grid-list-item/grid-list-item.component.ts:
```diff
@@ -227,6 +229,9 @@ export class GridListItemComponent<T> implements OnChanges, AfterViewInit, OnDes
     _index?: number;
 
+    /** @hidden */
+    _actionButtonClass = GRID_LIST_ITEM_ACTION_BUTTON_CLASS;
```
I haven't seen any rules about that.

@InnaAtanasova @droshev maybe you have some information about that?

InnaAtanasova replied:
We don't have a naming convention for the constants. I just checked and 95% of the constants have the same format. `GRID_LIST_ITEM_ACTION_BUTTON_CLASS` should be good, no need to prefix it with `_`.

Smiranin replied:
@InnaAtanasova I mean local constant `_actionButtonClass`. Should we rename it to `_GRID_LIST_ITEM_ACTION_BUTTON_CLASS`?

platon-rov commented on libs/platform/src/lib/components/approval-flow/approval-flow-team-list/approval-flow-team-list.component.scss:
```diff
@@ -1,6 +1,18 @@
+            @content;
+        }
+    }
+}
```
Agree with your suggestion. Also, I think we can have things like that project-wide, maybe we should implement them? @InnaAtanasova @droshev what do you think?

InnaAtanasova replied:
You can create a mixin in the current file to simplify your code but you can't create a mixin for the whole project as Fund-ngx is not supposed to have any sort of styling :) It's just consuming Fund-styles and adds custom css when angular specific changes are required, for examples, css that overwrites the css coming from CDK. The use of mixins is another reason we prefer to keep the css in fund-styles.

platon-rov replied:
https://github.com/SAP/fundamental-ngx/pull/6050#issuecomment-889075778

platon-rov commented on libs/platform/src/lib/components/approval-flow/approval-flow-node/approval-flow-node.component.ts:
```diff
@@ -182,35 +185,49 @@ export class ApprovalFlowNodeComponent implements OnInit, OnChanges, OnDestroy {
-        @Optional() private _rtlService: RtlService
+        private readonly _elRef: ElementRef,
+        private readonly _cdr: ChangeDetectorRef,
+        @Optional() private readonly _rtlService: RtlService
```
I guess it's the general rule, so it should be always set to `readonly`.

InnaAtanasova replied:
Created an issue for this: https://github.com/SAP/fundamental-ngx/issues/6121

Leave the `readonly` modifier for now, it's out of scope for this PR.

platon-rov replied:
Thank you

platon-rov commented on libs/platform/src/lib/components/approval-flow/approval-flow-node/approval-flow-node.component.scss:
```diff
@@ -362,210 +273,281 @@ $carousel-marker-top--edit-mode: $node-height--edit-mode / 2 - $carousel-marker-
+                    left: 0;
+                }
+            }
+        }
```
1. Fixed.
2. Yea, I guess there is no much sense to have styles for that component in the Fundamental Styles as we cannot provide that component with CSS only, it highly depends on the JS code. With Kevin, we agreed to leave styles for that component in the NGX. 

FYI @InnaAtanasova

InnaAtanasova replied:
The downside of leaving the scss code in fund-ngx is that this component won't be themable. The logic for changing the colors is entirely in fund-styles. In my opinion it's always to keep the css there, even if it's hidden from the official documentation. If the component doesn't work without JS nothing stops you from moving the css there but not creating a doc page for it. It's up to you and @KevinOkamoto but just keep in mind it won't be themable and most prob not gonna work in IE11.

platon-rov replied:
https://github.com/SAP/fundamental-ngx/pull/6050#issuecomment-889075778

platon-rov commented on libs/platform/src/lib/components/approval-flow/approval-flow-node/approval-flow-node.component.html:
```diff
@@ -120,27 +124,28 @@
+        [class.fdp-approval-flow-node__add--before]="!_blank && (placement === 'before' || placement === 'before-all')"
+        [class.fdp-approval-flow-node__add--before-all]="!_blank && placement === 'before-all'"
+        [class.fdp-approval-flow-node__add--after]="!_blank && (placement === 'after' || placement === 'after-all')"
+        [class.fdp-approval-flow-node__add--after-all]="placement === 'after-all'"
```
Oh, I also thought that it should be right way, but then found something 

https://github.com/angular/angular/issues/40623
https://angular.io/guide/built-in-directives#adding-and-removing-classes-with-ngclass

So they're recomeding to use `[class]` for multiple class binding, but not the `ngClass` directive. There is no explanation though.

platon-rov replied:
Oh, might be referenced https://hackmd.io/jzDc7hIDTdWtQblv2TbL9A

platon-rov replied:
Some more information for that particular case https://github.com/SAP/fundamental-ngx/pull/6050#discussion_r679214741

platon-rov commented on libs/core/src/lib/grid-list/components/grid-list-item/grid-list-item.component.html:
```diff
@@ -70,7 +70,7 @@
 
             <button
-                class="fd-grid-list__action-button"
+                [ngClass]="_actionButtonClass"
```
Here is some information https://github.com/SAP/fundamental-ngx/pull/6050#discussion_r678106147

InnaAtanasova replied:
While I see why you did it this way it's an approach that we have not yet agreed on. This will lead to people starting asking why it's done this way in one component and not in other. When we notice something that can be optimized we create an issue for this, set a quick meeting to discuss it and if agreed on using it we change the whole library. This will avoid having questions like the ones added by Andrii on the same PR regarding `readonly` :)

platon-rov replied:
Reverted, as `[ngClass]` is not a recommended approach anymore. 

Angular's `[class]` binding is not possible here, as `fd-button` has its own input field `class` which OK only with the strings, not with arrays/objects/etc.

InnaAtanasova commented on libs/platform/src/lib/components/approval-flow/approval-flow-messages/approval-flow-messages.component.ts:
```diff
@@ -0,0 +1,56 @@
+    }
+})
+export class ApprovalFlowMessagesComponent {
+    @Input()
+    messages: ApprovalFlowMessage[] = [];
+
+    @Input()
+    undoLastActionAvailable = false;
+
+    @Output()
+    messagesChange = new EventEmitter<ApprovalFlowMessage[]>();
+
+    @Output()
+    undoLastAction = new EventEmitter<void>();
```
Typedocs?

InnaAtanasova commented on libs/platform/src/lib/components/approval-flow/approval-flow-node/approval-flow-drop-zone.directive.ts:
```diff
@@ -1,22 +1,28 @@
 export class ApprovalFlowDropZoneDirective {
-    @Input() placement: 'before'| 'after' | 'before-all' | 'after-all';
+    @Input()
+    placement: ApprovalFlowNewNodePlacement;
 
     active = false;
 
-    constructor(private elRef: ElementRef) {}
+    constructor(
+        private readonly elRef: ElementRef
+    ) {}
```
typedocs?

InnaAtanasova commented on libs/platform/src/lib/components/approval-flow/approval-flow-node/approval-flow-node.component.html:
```diff
@@ -1,45 +1,46 @@
+</ng-template>
+
+<ng-template #nodeContent>
+    <div class="fdp-approval-flow-node__inner">
```
I'm not sure about the `fdp-` prefix. This will trigger a lot of questions from the users side.

platon-rov replied:
Why not? 

We have `fd-*` prefix for all the stuff related to the Fundamental-NGX Core. 
And we're actually have the `fdp-*` prefix in a lot of Platform components.

InnaAtanasova commented on libs/platform/src/lib/components/approval-flow/approval-flow-add-node/approval-flow-add-node.component.scss:
```diff
@@ -1,8 +1,8 @@
 .#{$block} {
-    &__dialog {
-        &-body {
+    .fd-dialog {
```
can be directly `.fd-dialog__body` :)

InnaAtanasova commented on libs/platform/src/lib/components/approval-flow/approval-flow-add-node/approval-flow-add-node.component.scss:
```diff
@@ -1,8 +1,8 @@
-        &-body {
+    .fd-dialog {
+        &__body {
             width: 320px;
             height: 433px;
```
maybe in rem?

InnaAtanasova reviewed (commented):
I really advise you to move the css in Fundamental-styles. Like I mentioned in one of my comments, if the component really requires a lot of JS to display it in the docs, you can not show it in the docs. But I see no problem moving the code there. You won't need the variables fallback, you will make use of lots of mixins and most importantly you will have theming support. Keeping the code here is hard to maintain.

platon-rov commented:
> I really advise you to move the css in Fundamental-styles. Like I mentioned in one of my comments, if the component really requires a lot of JS to display it in the docs, you can not show it in the docs. But I see no problem moving the code there. You won't need the variables fallback, you will make use of lots of mixins and most importantly you will have theming support. Keeping the code here is hard to maintain.

Okay, sure, so I've created #6123. Currently designs do not fully cover Fiori3 themes, so I've notified Keny Kim (designer) about it. I think once there will be full themes support we can start refactor CSS code.

Anyway that PR still has its value as many things got refactored for better code quality.

mikerodonnell89 commented:
- [ ] I feel like the + circles could be 1px higher
<img width="73" alt="Screen Shot 2021-08-03 at 2 21 35 PM" src="https://user-images.githubusercontent.com/2471874/128066393-86e10a4c-cf47-43e1-86ce-82522f19762b.png">

- [ ] There is too much padding here in RTL mode:
<img width="431" alt="Screen Shot 2021-08-03 at 2 22 56 PM" src="https://user-images.githubusercontent.com/2471874/128066692-7714206c-25f8-4728-97fe-a49b7a19053f.png">

- [ ] After clicking the checkbox to select someone the cursor changes from pointer to default, while still hovered over the checkbox

platon-rov commented:
> * [ ]  I feel like the + circles could be 1px higher

Fixed

> * [ ]  There is too much padding here in RTL mode:

Fixed

> * [ ]  After clicking the checkbox to select someone the cursor changes from pointer to default, while still hovered over the checkbox

It's derived from the checkbox component (radio has the same behavior), should it be changed?
